e2e_playwright/hostframe_app_test.py

```diff
@@ -0,0 +1,187 @@
+    expect(menu_list.get_by_text("Adopt a Corgi")).to_be_attached()
+
+
+def test_handles_host_toolbar_item_message(
+    iframed_app: IframedPage, assert_snapshot: ImageCompareFunction
+):
+    frame_locator, toolbar_buttons = _load_html_and_get_locators(iframed_app)
+    toolbar_buttons.get_by_text("Add Toolbar Item").click()
+
+    toolbar_action_button = frame_locator.get_by_test_id("stToolbarActionButton")
+    expect(toolbar_action_button).to_have_count(2)
+    expect(toolbar_action_button.nth(0)).to_have_text("Favorite")
+    expect(toolbar_action_button.nth(1)).to_have_text("Share")
+
+    assert_snapshot(
+        frame_locator.get_by_test_id("stApp"), name="hostframe_app-toolbar_items"
+    )
+
+
+def test_handles_hide_sidebar_nav_message(iframed_app: IframedPage):
+    frame_locator, toolbar_buttons = _load_html_and_get_locators(iframed_app)
+    expect(frame_locator.get_by_test_id("stSidebarNav")).to_be_attached()
+    toolbar_buttons.get_by_text("Hide Sidebar Nav").click()
+    expect(frame_locator.get_by_test_id("stSidebarNav")).not_to_be_attached()
+
+
+def test_handles_sidebar_downshift_message(iframed_app: IframedPage):
+    frame_locator, toolbar_buttons = _load_html_and_get_locators(iframed_app)
+    # Trigger sidebar downshift
+    toolbar_buttons.get_by_text("Sidebar Chevron Downshift").click()
+    # Hover over sidebar to reveal the button
+    frame_locator.get_by_test_id("stSidebarContent").hover()
+    # Close the sidebar
+    frame_locator.get_by_test_id("stSidebar").locator("button").click()
+    # Check chevron positioning
+    expect(frame_locator.get_by_test_id("stSidebarCollapsedControl")).to_have_css(
+        "top", "50px"
+    )
+
+
+def test_handles_host_terminate_and_restart_websocket_connection_messages(
+    iframed_app: IframedPage,
+):
+    frame_locator, toolbar_buttons = _load_html_and_get_locators(iframed_app)
+
+    # Kill the websocket connection and verify that the app moves into an
+    # error state.
+    toolbar_buttons.get_by_text("Terminate Websocket").click()
+    expect(frame_locator.get_by_test_id("stStatusWidget")).to_have_text("Error")
+
+    frame = frame_locator.owner.page.frame("guest")
+    assert frame is not None
+    # start observing our connection statuses before we click on restart websocket
+    register_connection_status_observer(frame)
+
+    # Request that the websocket connection gets restarted.
+    toolbar_buttons.get_by_text("Restart Websocket").click()
+    wait_until(
+        iframed_app.page,
+        lambda: len(get_observed_connection_statuses(frame)) == 3,
+        timeout=5000,
+    )
+    statuses = get_observed_connection_statuses(frame)
+    assert statuses[0] == "PINGING_SERVER"
+    assert statuses[1] == "CONNECTING"
+    assert statuses[2] == "CONNECTED"
+
+    # Check that the status widget disappeared again as the app reruns and completes.
+    expect(frame_locator.get_by_test_id("stStatusWidget")).not_to_be_attached()
```

Comment:
nit: alternative here might be to expect the `data-test-script-state`  attribute of `stApp` to be `notRunning`

Reply:
updated it!